app/src/main/java/ch/epfl/sdp/HistoryFragment.java

```diff
@@ -57,44 +61,31 @@ public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
     }
 
-    @VisibleForTesting
-    void setFirestoreInteractor(FirestoreInteractor interactor) {
-        db.setFirestoreInteractor(interactor);
-    }
-
-    private void initQueryHandler() {
-        ArrayAdapter historyAdapter = new ArrayAdapter<>(view.getContext(), android.R.layout.simple_list_item_1);
-        ListView historyTracker = view.findViewById(R.id.history_tracker);
-        historyTracker.setAdapter(historyAdapter);
-
-        handler = new QueryHandler<QuerySnapshot>() {
-            @Override
-            public void onSuccess(QuerySnapshot snapshot) {
-                historyAdapter.clear();
-                connectionStatus.setText(R.string.query_ok);
-                for (QueryDocumentSnapshot qs : snapshot) {
-                    try {
-                        Map<String, Object> positionRecord = qs.getData();
-                        historyAdapter.insert(String.format("Found @ %s:%s on %s",
-                                ((GeoPoint)(positionRecord.get("geoPoint"))).getLatitude(),
-                                ((GeoPoint)(positionRecord.get("geoPoint"))).getLongitude(),
-                                ((Timestamp)(positionRecord.get("timeStamp"))).toDate()), 0);
-                    } catch (NullPointerException e) {
-                        historyAdapter.insert("[...unreadable.:).]", 0);
-                    }
+    private CompletableFuture<Void> refreshHistory(View view) {
+        connectionStatus.setText(R.string.loading_with_dots);
+        return db.readHistory().thenAccept(res -> {
+            ArrayAdapter historyAdapter = new ArrayAdapter<>(view.getContext(), android.R.layout.simple_list_item_1);
+            ListView historyTracker = view.findViewById(R.id.history_tracker);
+            historyTracker.setAdapter(historyAdapter);
+            historyAdapter.clear();
+
+            connectionStatus.setText(R.string.query_ok);
+
+            for (Map.Entry<String, Map<String, Object>> mapEntry : res.entrySet()) {
+                try {
+                    Map<String, Object> positionRecord = mapEntry.getValue();
+                    historyAdapter.insert(String.format("Found @ %s:%s on %s",
```

Comment:
We can use resource strings with interpolation to remove this hardcoded string, but that is out of the scope for this PR

Reply:
I left a TODO for next PR